File: js/level/level.js

// var d_lastX = 0;
// var d_lastY = 0;
// var d_once = 0;

Level = Class.extend({
	entities: new Array(),
	w: 0,
	h: 0,
	map: null,
	towns: null,
	seed: null,
	minimap: null,

	init: function(w, h, seed) {
		this.w = w;
		this.h = h;
		this.seed = seed;

		Math.seedrandom(seed);

		console.log("Creating level with seed: " + seed);

		var levelGen = new LevelGen();
		this.map = levelGen.createSurfaceMap(w, h);
		this.towns = levelGen.getTowns();

		// for (var y = 0; y < w; y++) {
		// 	for (var x = 0; x < h; x++) {
		// 		console.log("this.map["+y+"]["+x+"]="+this.map[y][x]);
		// 	}
		// }
	},

	addEntity: function(entity) {
		this.entities.push(entity);
	},

	renderBackground: function(screen, xScroll, yScroll) {
		var x0 = xScroll >> 5;
		var y0 = yScroll >> 5;

		screen.setOffset(xScroll, yScroll);
		for (var y = y0; y < screen.yTiles + y0 + 2; y++) {
			for (var x = x0; x < screen.xTiles + x0 + 2; x++) {
				if (y < 0 || x < 0 || y >= this.h || x >= this.w) {
					gTileLibrary['water'].render(screen, x * screen.tileSize, y * screen.tileSize);
				} else {
					var townTile = false;
					for (var i = 0; i < this.towns.length; i++) {
						if (this.towns[i].isInTown(x, y, 0)) {
							townTile = true;
							break;
						}
					}
					var currentTile = gTileLibrary[this.map[y][x].id];
					if (townTile && currentTile.mayPass(this, x, y)) {
						gTileLibrary['dirt'].render(screen, x * screen.tileSize, y * screen.tileSize);
					} else {
						currentTile.render(screen, x * screen.tileSize, y * screen.tileSize);
					}
				}
			}
		}
		screen.setOffset(0, 0);
	},

	generateMinimap: function(screen) {
		this.minimap = screen.ctx.createImageData(this.w, this.h);
		var color = { r: 0, g: 0, b: 0, a: 255 };
		for (var y = 0; y < this.h; y++) {
			for (var x = 0; x < this.w; x++) {
				var index = (x + y * this.minimap.width) * 4;

				if (gTileLibrary[this.map[y][x].id].name == 'grass') {
					color.r = 100;
					color.g = 164;
					color.b = 44;
				}
				if (gTileLibrary[this.map[y][x].id].name == 'water') {
					color.r = 131;
					color.g = 198;
					color.b = 255;
				}
				if (gTileLibrary[this.map[y][x].id].name == 'sand') {
					color.r = 217;
					color.g = 183;
					color.b = 92;
				}
				if (gTileLibrary[this.map[y][x].id].name == 'rock') {
					color.r = 129;
					color.g = 139;
					color.b = 139;
				}
				if (gTileLibrary[this.map[y][x].id].name == 'tree') {
					color.r = 11;
					color.g = 92;
					color.b = 47;
				}
				if (gTileLibrary[this.map[y][x].id].name == 'cactus') {
					color.r = 129;
					color.g = 161;
					color.b = 0;
				}

				this.minimap.data[index + 0] = color.r;
				this.minimap.data[index + 1] = color.g;
				this.minimap.data[index + 2] = color.b;
				this.minimap.data[index + 3] = color.a;
			}
		}
	},

	renderMinimap: function(screen, xScroll, yScroll) {
		if (this.minimap == null) {
			this.generateMinimap(screen);
		}

		screen.ctx.putImageData(this.minimap, 0, 0);
	},

	renderSprites: function(screen, xScroll, yScroll, player) {
		for (var i = 0; i < this.entities.length; i++) {
			this.entities[i].render(screen, xScroll, yScroll);
		}
		for (var i = 0; i < this.towns.length; i++) {
			this.towns[i].render(screen, player.x, player.y);
		}
	},

	tick: function(msPerTick) {
		for (var i = 0; i < this.entities.length; i++) {
			this.entities[i].tick(msPerTick);
		}
		for (var i = 0; i < this.towns.length; i++) {
			this.towns[i].tick(msPerTick);
		}
	},

	getTile: function(xt, yt) {
		// if (d_lastX != xt || d_lastY != yt) {
		// 	d_lastX = xt;
		// 	d_lastY = yt;

		// 	console.log("tile @ xt="+xt+" /yt="+yt+" tile:"+this.map[yt][xt]);
		// }
		return gTileLibrary[this.map[yt][xt].id];
	},

	bumpedInto: function(level, xt, yt, entity) {
		for (var i = 0; i < this.towns.length; i++) {
			if (this.towns[i].isInTown(xt, yt, 0)) {
				this.towns[i].enter();
			} else {
				this.towns[i].leave();
			}
		}
	},
});

File: src/xrpg/game/level/Level.java
package xrpg.game.level;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

import xrpg.game.Screen;
import xrpg.game.entity.Entity;
import xrpg.game.entity.Player;
import xrpg.game.level.tile.Tile;

public class Level {
	public Random m_random;
	public String m_seed;
	public Player m_player;
	
	public int m_w;
	public int m_h;
	
	public char[][] m_map;
	public List<Town> m_towns;
	public List<Entity> m_entities = new ArrayList<Entity>();
	
	public Level(int width, int height, String seed) {
		m_seed = seed;
		m_random = new Random(seed.hashCode());
		
		m_w = width;
		m_h = height;
		
		System.out.println("Creating level with sed: " + seed);

		m_map = LevelGen.createSurfaceMap(m_random, m_w, m_h);
		m_towns = LevelGen.getTowns(m_random, m_w, m_h, m_map);

		// for (int y = 0; y < m_w; y++) {
		// 	for (int x = 0; x < m_h; x++) {
		// 		System.out.println("this.map["+y+"]["+x+"]="+this.map[y][x]);
		// 	}
		// }
	}
	
	public boolean addEntity(Entity entity) {
		return m_entities.add(entity);
	}
	
	public void renderBackground(Screen screen, int xScroll, int yScroll) throws IOException {
		int x0 = xScroll >> 5;
		int y0 = yScroll >> 5;

		screen.setOffset(xScroll, yScroll);
		for (int y = y0; y < Screen.YTILES + y0 + 2; y++) {
			for (int x = x0; x < Screen.XTILES + x0 + 2; x++) {
				if (y < 0 || x < 0 || y >= m_h || x >= m_w) {
					Tile.ms_water.render(screen, x * Screen.TILESIZE, y * Screen.TILESIZE);
				} else {
					boolean townTile = false;
					for (Town town : m_towns) {
						if (town.isInTown(x, y, 0)) {
							townTile = true;
							break;
						}
					}
					Tile currentTile = Tile.getTile(m_map[y][x]);
					if (townTile && currentTile.mayPass(this, x, y, null)) {
						Tile.ms_dirt.render(screen, x * Screen.TILESIZE, y * Screen.TILESIZE);
					} else {
						currentTile.render(screen, x * Screen.TILESIZE, y * Screen.TILESIZE);
					}
				}
			}
		}

		screen.setOffset(0, 0);
	}

	public void generateMinimap(Screen screen) {
		// todo
	}
	
	public void renderMinimap(Screen screen, int xScroll, int yScroll) {
		// todo
	}
	
	public void renderSprites(Screen screen, int xScroll, int yScroll) throws IOException {
		for (Entity entity : m_entities) {
			entity.render(screen, xScroll, yScroll);
		}
		for (Town town : m_towns) {
			town.render(screen, m_player.m_x, m_player.m_y);
		}
	}
	
	public void tick() {
		for (int i = 0; i < m_entities.size(); i++) {
			Entity entity = m_entities.get(i);
			entity.tick();
			
			if (entity.m_remove) {
				m_entities.remove(i--);
			}
		}
		for (Town town : m_towns) {
			town.tick();
		}
	}
	
	public Tile getTile(int xt, int yt) {
		return Tile.getTile(m_map[yt][xt]);
	}

	public void bumpedInto(int xt, int yt, Entity entity) {
		if (Tile.getTile(m_map[yt][xt]).bumpedInto(this, xt, yt, entity) == false) {
			return;
		}

		for (Town town : m_towns) {
			if (town.isInTown(xt, yt, 0)) {
				town.enter();
			} else {
				town.leave();
			}
		}
	}
}
